fcos translate.go: add warn on small or constrained root partition

prestist opened the pull request:
https://github.com/coreos/butane/issues/211



Note I somehow dropped the '1' in the version for FCOS in the commit messages.. but before I update the commit structure lets make sure we are happy with the functional changes. Look at the first 2 commits, i.e the exp changes this is basically propagated to all versions back to 1_3 as you can see later with the fixup commits. Once approved I will merge the fixup commits with a rename to the commits to be inclusive of the changes!

bgilbert commented on config/fcos/v1_5_exp/translate.go:
```diff
@@ -68,17 +68,31 @@ func (c Config) ToIgn3_4Unvalidated(options common.TranslateOptions) (types.Conf
+						r.AddOnWarn(path.New("json", "storage", "disks", i, "partitions", p, "sizeMiB", *partition.SizeMiB), common.ErrRootTooSmall)
+					} else {
+						for _, op := range disk.Partitions {
+							if op.Number > partition.Number {
```
We're currently performing this check for every partition after the root partition, not just the one immediately after it.

prestist replied:
OH can we assume that they are always in order?

bgilbert replied:
What we don't want is to warn repeatedly for every additional partition.  But you're right, we can't assume they're in order, either by starting offset or by partition number.  And actually, we can't make assumptions about `op.Number` either, since it might be 0.

So we'll need an heuristic for figuring out the earliest partition start.  If the `startMiB` is always specified, this is easy.  Otherwise, the partition order is determined by a combination of `sgdisk`'s heuristics (see the man page for the `-n` option) and the additional heuristics applied by [partitions.go](https://github.com/coreos/ignition/blob/main/internal/exec/stages/disks/partitions.go).  And unfortunately, they have some information we don't, which is the overall size of the disk, and thus the size of the largest available block.

So we'll need to figure this out.  If there are cases where we're not sure, it's better to not warn.

bgilbert commented on config/fcos/v1_5_exp/translate.go:
```diff
@@ -68,17 +68,31 @@ func (c Config) ToIgn3_4Unvalidated(options common.TranslateOptions) (types.Conf
+					} else {
+						for _, op := range disk.Partitions {
+							if op.Number > partition.Number {
+								if op.StartMiB == nil {
```
This doesn't match the logic described in https://github.com/coreos/butane/issues/211#issuecomment-1205967355.

bgilbert commented on config/common/errors.go:
```diff
@@ -37,6 +37,8 @@ var (
 	ErrNoFilesDir             = errors.New("local file paths are relative to a files directory that must be specified with -d/--files-dir")
 	ErrTreeNotDirectory       = errors.New("root of tree must be a directory")
+	ErrRootTooSmall           = errors.New("root should have 8GiB of space available")
+	ErrRootNotLastPartition   = errors.New("root should be last partition number to allow for growth")
```
The message is misleading, since we allow and encourage adding partitions after the root.

bgilbert commented on config/common/errors.go:
```diff
@@ -37,6 +37,8 @@ var (
 	ErrNodeExists             = errors.New("matching filesystem node has existing contents or different type")
 	ErrNoFilesDir             = errors.New("local file paths are relative to a files directory that must be specified with -d/--files-dir")
 	ErrTreeNotDirectory       = errors.New("root of tree must be a directory")
+	ErrRootTooSmall           = errors.New("root should have 8GiB of space available")
```
nit: `available` could be read to imply free space in the filesystem.

bgilbert commented on config/fcos/v1_5_exp/translate.go:
```diff
@@ -68,17 +68,31 @@ func (c Config) ToIgn3_4Unvalidated(options common.TranslateOptions) (types.Conf
-				if partition.Label != nil {
+		for p, partition := range disk.Partitions {
+			if partition.Label != nil {
+				if *partition.Label == "root" {
```
Unnecessary nesting.

bgilbert commented on config/fcos/v1_5_exp/translate.go:
```diff
@@ -68,17 +68,31 @@ func (c Config) ToIgn3_4Unvalidated(options common.TranslateOptions) (types.Conf
+						}
+					}
+				}
+				// Don't warn if wipeTable is set, matching later spec versions
```
Looks like we lost the original comment here.  "matching later spec versions"?

bgilbert commented on config/fcos/v1_5_exp/translate_test.go:
```diff
@@ -146,6 +146,198 @@ func TestTranslateBootDevice(t *testing.T) {
+					{
+						Kind:    report.Warn,
+						Message: common.ErrRootTooSmall.Error(),
+						Context: path.New("json", "storage", "disks", 0, "partitions", 0, "sizeMiB", 500),
```
This is specifying the 501st element of the `sizeMiB` array.

bgilbert commented on config/fcos/v1_5_exp/translate_test.go:
```diff
@@ -146,6 +146,198 @@ func TestTranslateBootDevice(t *testing.T) {
+							Partitions: []types.Partition{
+								{
+									Label:   util.StrToPtr("root"),
+									SizeMiB: util.IntToPtr(10000),
```
This size is sufficient, so the test should pass.

bgilbert commented on config/fcos/v1_5_exp/translate_test.go:
```diff
@@ -146,6 +146,198 @@ func TestTranslateBootDevice(t *testing.T) {
+					},
+				},
+			},
+		},
```
Let's also add tests where the root partition doesn't explicitly specify a size but is constrained by another partition which does/doesn't leave enough room.

bgilbert commented on config/fcos/v1_5_exp/translate_test.go:
```diff
@@ -146,6 +146,198 @@ func TestTranslateBootDevice(t *testing.T) {
+					{
+						Kind:    report.Warn,
+						Message: common.ErrRootNotLastPartition.Error(),
+						Context: path.New("json", "storage", "disks", 0, "partitions", 0, "number", 4),
```
Likewise, this is the 5th element of the `number` array.

prestist committed:
fcos translate.go: add warn on small or constrained root partition

Fixes: #211, the root partition needs a certain amount of space, with
recent changes butane has the opportunity to warn when those expectations
are not met.

Warn if root partition sizeMiB is explicitly less than 8GiB, or if root
is set to fill available space and any other partition on the disk is
auto-positioned (startMiB == 0 or nil), which would prevent root from
expanding.

Applied to FCOS versions v1_3 through v1_8_exp.

prestist committed:
fcos translate_test: add tests for partition constraint detection

Add test cases for:
 - Root partition constrained by disk position (not partition number)
 - Root partition with auto-positioned following partition (StartMiB=0)
 - Root partition NOT constrained when next partition has explicit StartMiB
 - Root partition constrained by auto-positioned partition even when an
   explicit partition is also present
 - Root partition size validation (too small, exactly 8GiB)

Applied to FCOS versions v1_3 through v1_8_exp.

aaradhak commented:
The first commit message says fcos v0_7_exp but there is no v0_7_exp directory. May be this should be fcos v1_7_exp?

aaradhak commented:
Minor: Typo in commit message body: "contstraints" should be "constraints"

aaradhak commented on config/common/errors.go:
```diff
@@ -68,6 +68,8 @@ var (
 	ErrReuseByLabel         = errors.New("partitions cannot be reused by label; number must be specified except on boot disk (/dev/disk/by-id/coreos-boot-disk) or when wipe_table is true")
 	ErrWrongPartitionNumber = errors.New("incorrect partition number; a new partition will be created using reserved label")
+	ErrRootTooSmall         = errors.New("root should have 8GiB of space assigned")
+	ErrRootConstrained      = errors.New("root is configured too small, and has no room to expand")
```
just curious if this can this be rephrased? shouldnt this error message say something like that it's configured to fill available space?

prestist replied:
Good point! The current message "root is configured too small, and has no room to expand"
  could be clearer.

aaradhak commented on config/fcos/v1_7/translate.go:
```diff
@@ -71,16 +71,70 @@ func (c Config) ToIgn3_6Unvalidated(options common.TranslateOptions) (types.Conf
+
+				// In the boot_device.mirror case, nothing specifies partition numbers
+				// so match existing partitions only when `wipeTable` is false
+				if !util.IsTrue(disk.WipeTable) {
```
Trying to understand this wipetable change. The previous code only checked reserved partlabels when wipeTable is false. The current change now runs the root validation even when wipeTable is true. I guess it is intentional.

prestist replied:
Yeah it was intentional the root partition size validation needs to happen regardless of wipeTable, because:

  1. Even when wipeTable is true (creating a fresh partition table), users can still specify
     partitions that trap the root partition - this is exactly the scenario we want to warn about

  2. The wipeTable check only applies to the "reserved partlabel validation" (checking that
     BIOS-BOOT is partition 1, root is partition 4, etc.). That validation should only run
     when wipeTable is false because we're matching existing partitions.

  3. The root size/constraint validation is orthogonal to whether we're wiping the table -
     it's about warning users before they create a configuration that won't have enough space
     for OS updates.

prestist commented:
@aaradhak  

Fixed both issues! I've rebased and updated the commit messages:

  1. Changed from "fcos v0_7_exp" to "fcos translate.go" / "fcos translate_test.go" since
     these changes actually apply to FCOS versions v1_3 through v1_7

  2. Fixed the typo: "contstraints" → "constraints"

aaradhak commented on config/fcos/v1_7/translate.go:
```diff
@@ -71,16 +71,70 @@ func (c Config) ToIgn3_6Unvalidated(options common.TranslateOptions) (types.Conf
+										nextStart = *nextPartition.StartMiB
+									}
+									// If this partition starts before the current "next", it's closer to root
+									if (apStart > 0 && apStart < nextStart) || (nextStart == 0 && apStart > 0) {
```
The condition here causes an explicitly positioned partition to replace an auto-positioned one as the "nearest" candidate but the auto-positioned partition is the one that actually constrains root right?

prestist replied:
oooof, okay; yees you are right; Let me update the code to account for this.

prestist commented:
@aaradhak Good catch! You're right - the previous logic missed when both an auto-positioned and an explicitly positioned partition were present, the "find nearest"  would prefer the explicit partitions, effectively hiding the auto-positioned partition that actually constrains root.                                                                                          
                                                                                                                                                                         
The startMiB of any following partition. If this is unspecified (0) and the root sizeMiB is also unspecified, then the following partition will prevent the root partition from growing at all, and we should warn.

Instead of trying to determine which partition is physically "nearest" to root (which we can't reliably do without knowing disk size and sgdisk's placement heuristics), we now simply check: if root is auto-sized AND any other partition on the disk has startMiB unspecified, warn. I also added a test case for this scenario (auto-positioned partition with an explicit one also present).

aaradhak reviewed (approved):
LGTM
